build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/StringTemplateTask.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.internal;
+
+import org.gradle.api.Action;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.GradleException;
+import org.gradle.api.file.DirectoryProperty;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.api.tasks.TaskAction;
+import org.stringtemplate.v4.ST;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public abstract class StringTemplateTask extends DefaultTask {
+
+    private final ListProperty<TemplateSpec> templateSpecListProperty;
+    private final DirectoryProperty outputFolder;
+
+    @Inject
+    public StringTemplateTask(ObjectFactory objectFactory) {
+        templateSpecListProperty = objectFactory.listProperty(TemplateSpec.class);
+        outputFolder = objectFactory.directoryProperty();
+    }
+
+    public void template(Action<TemplateSpec> spec) {
+        TemplateSpec templateSpec = new TemplateSpec();
+        spec.execute(templateSpec);
+        templateSpecListProperty.add(templateSpec);
+    }
+
+    @Nested
+    public ListProperty<TemplateSpec> getTemplates() {
+        return templateSpecListProperty;
+    }
+
+    @OutputDirectory
+    public DirectoryProperty getOutputFolder() {
+        return outputFolder;
+    }
+
+    @TaskAction
+    public void generate() {
+        File outputRootFolder = getOutputFolder().getAsFile().get();
```

Comment:
Should we clear out the output folder first? Let's say you delete a template spec, what's going to remove that orphaned file?

Reply:
++. Done

Reply:
I would rather not have the plugin do deletions. The plugin has no notion of what else may be in the output folder.

Reply:
There should be nothing in that output folder that wasn't placed there by that task. This is a strict assumption by Gradle. If we expect to have multiple source generation plugins, they need to have unique output directories. In fact, having multiple tasks share an output folder will be reported as an error by Gradle.

Reply:
The only generated output in that directory is from this Gradle task.

I have found it useful to keep local copies of things around for diff purposes. If this task were to delete such, then that would not be possible. (I would have to otherwise copy the whole directory)

Reply:
Right, but what happens if you remove a template? The output will remain and the only solution is to run `clean`, which then breaks incremental builds. The task should either a) clean the output directory on each execution to ensure there are no output there that aren't created by this task or b) implement some logic to explicitly delete superfluous output. In general, it should never be _necessary_ to run `clean`. Tasks should be incremental and adapt to any change in input and produce the right output. This is no different than if I delete a Java source file and run compile, the corresponding class file should be removed from the output directory (which is the Gradle behavior).

If you want to compare changes then yes, you'll need to copy the output somewhere else temporarily. This may be more inconvenient but is more _correct_. Perhaps you can describe such a use case in detail.

Reply:
When writing a template it is most straightforward to write the Java source code, then extract the template from it. The primary concern is with somewhat automated refreshes of the project when in the IDE, that those my inadvertently delete the handwritten code before the template has been created from it. I guess one could just do this sort of work in a different directory.

Reply:
Cleanup output dir done, see [2ea06f1](https://github.com/elastic/elasticsearch/pull/97388/commits/2ea06f1bd560edba62ab208f6065a0d6090a4945)

Reply:
Right, I'd just write the hand-written source code outside the generated-src directory, then you don't have to worry about it getting clobbered.